cmd/abapEnvironmentCreateTag.go

```diff
@@ -0,0 +1,246 @@
+		err = createTagsForSingleItem(item, telemetryData, connection, client, com)
+		if err != nil {
+			errorOccurred = true
+		}
+	}
+
+	if errorOccurred {
+		message := "At least one tag has not been created"
+		log.Entry().Errorf(message)
+		return errors.New(message)
+	}
+	return nil
+
+}
+
+func createTagsForSingleItem(item CreateTagBacklog, telemetryData *telemetry.CustomData, con abaputils.ConnectionDetailsHTTP, client piperhttp.Sender, com abaputils.Communication) (err error) {
+
+	for index := range item.tags {
+		err = createSingleTag(item, index, telemetryData, con, client, com)
+	}
+	return err
+}
+
+func createSingleTag(item CreateTagBacklog, index int, telemetryData *telemetry.CustomData, con abaputils.ConnectionDetailsHTTP, client piperhttp.Sender, com abaputils.Communication) (err error) {
+
+	requestBodyStruct := CreateTagBody{RepositoryName: item.repositoryName, CommitID: item.commitID, Tag: item.tags[index].tagName, Description: item.tags[index].tagDescription}
+	errorMessage := "Could not create tag " + requestBodyStruct.Tag + " for repository " + requestBodyStruct.RepositoryName + " with commitID " + requestBodyStruct.CommitID
+	requestBodyJson, err := json.Marshal(&requestBodyStruct)
+	if err != nil {
+		return err
+	}
+
+	log.Entry().Debugf("Request body: %s", requestBodyJson)
+	resp, err := abaputils.GetHTTPResponse("POST", con, requestBodyJson, client)
+	if err != nil {
+		err = abaputils.HandleHTTPError(resp, err, errorMessage, con)
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Parse response
+	var body CreateTagResponse
+	var abapResp map[string]*json.RawMessage
+	bodyText, _ := ioutil.ReadAll(resp.Body)
+
+	json.Unmarshal(bodyText, &abapResp)
+	json.Unmarshal(*abapResp["d"], &body)
+
+	con.URL = con.Host + "/sap/opu/odata/sap/MANAGE_GIT_REPOSITORY/Pull(guid'" + body.UUID + "')"
+	err = checkStatus(con, client, com)
+
+	if err == nil {
+		log.Entry().Info("Created tag " + requestBodyStruct.Tag + " for repository " + requestBodyStruct.RepositoryName + " with commitID " + requestBodyStruct.CommitID)
+	} else {
+		log.Entry().Error("NOT created: Tag " + requestBodyStruct.Tag + " for repository " + requestBodyStruct.RepositoryName + " with commitID " + requestBodyStruct.CommitID)
+	}
+
+	return err
+}
+
+func checkStatus(con abaputils.ConnectionDetailsHTTP, client piperhttp.Sender, com abaputils.Communication) (err error) {
+	status := "R"
+	pollIntervall := com.GetPollIntervall()
+	count := 0
+	for {
+		count += 1
+		entity, _, err := abaputils.GetStatus("Could not create Tag", con, client)
+		if err != nil {
+			return err
+		}
+		status = entity.Status
+		if status != "R" {
+			if status == "E" {
+				err = errors.New("Could not create Tag")
+			}
+			return err
+		}
+		if count >= 200 {
+			return errors.New("Could not create Tag (Timeout)")
+		}
+		time.Sleep(pollIntervall)
+	}
+}
+
+func prepareBacklog(config *abapEnvironmentCreateTagOptions) (backlog []CreateTagBacklog, err error) {
```

Comment:
I do not know the specification so I can only guess but see my comment for the documentation file... for me the logic (if I understand which I'm not sure) is a little strange
Wouldn't it make sense to have EITHER the addondescriptor config file OR the manual specified repository? why is addondescriptor config file mandatory and why are there side effects of the different options which most likely no end user would ever like to have?

Reply:
Yes, I'm still thinking about that. In other steps, both options are available at the same time and the configuration is merged...

Reply:
it is now disallowed to specify repositoryName && repositories at the same time